### PR TITLE
Add check phase to ninja setup hook

### PR DESCRIPTION
**Pull request by matthewbauer**
###### Motivation for this change

Most tests in Ninja packages were not getting run because the default check phase assumes you have a Makefile. This adds a check phase to Ninja's setup-hook to fix this. Here is what I changed:

- Respect installFlags
- Automatically add checkPhase (can be disabled with dontUseNinjaCheck
  in the same way as dontUseNinjaBuild and dontUseNinjaInstall). Tests
  are only run when "ninja test" exists.
- Error in build phase when build.ninja is missing. We don’t have a
  way to fall back to other build methods, so it’s best to be very
  clear when we aren’t able to build with ninja
- Set -l flag to 1 when enableParallelBuilding is disabled

Intended as an alternative to #50323.

### CODE REVIEW

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh [hunk 1]
```diff
@@ -48,3 +45,40 @@ ninjaInstallPhase() {
...
+    runHook preCheck
+
+    if [ -z "${checkTarget:-}" ]; then
+        if ninja -n test >/dev/null 2>&1; then
+            checkTarget=test
+        fi
+    fi
+
+    if [ -z "${checkTarget:-}" ]; then
+        echo "no check/test target in ${makefile:-Makefile}, doing nothing"
```
Comment (jtojnar) [inline pkgs/development/tools/build-managers/ninja/setup-hook.sh]: makefile?

Comment (matthewbauer) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: Thanks!

**Comment by matthewbauer** [comment]
@jtojnar This looks like it breaks a few packages that have broken tests. Especially with meson+ninja packages (gnome-keyring). 

/cc @oxij

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh [hunk 2]
```diff
@@ -32,11 +28,12 @@ fi
...
     runHook preInstall
 
-    installTargets="${installTargets:-install}"
-
     # shellcheck disable=SC2086
-    local flagsArray=( $installTargets \
-        $ninjaFlags "${ninjaFlagsArray[@]}")
+    local flagsArray=(
+        $ninjaFlags "${ninjaFlagsArray[@]}"
+        $installFlags "${installFlagsArray[@]}"
```
Comment (orivej) [inline pkgs/development/tools/build-managers/ninja/setup-hook.sh]: Ninja does not have any flags that would be useful with the install target. I'd revert the change to `glib/setup-hook.sh` and remove this line, as it may break other setup hooks too.

**Review by orivej — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh [hunk 3]
```diff
@@ -1,26 +1,22 @@
...
+        buildCores="$NIX_BUILD_CORES"
+    fi
 
-        # shellcheck disable=SC2086
-        local flagsArray=( \
-            -j"$buildCores" -l"$NIX_BUILD_CORES" \
-            $ninjaFlags "${ninjaFlagsArray[@]}" \
-            $buildFlags "${buildFlagsArray[@]}")
+    local flagsArray=(
+        -j$buildCores -l$buildCores
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: I am hoping it has the same behavior as in make? The only difference is it is autodetected when left out in ninja.

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh [hunk 4]
```diff
@@ -48,3 +45,40 @@ ninjaInstallPhase() {
...
+        echoCmd 'check flags' "${flagsArray[@]}"
+        ninja "${flagsArray[@]}"
+        unset flagsArray
+    fi
+
+    runHook postCheck
+}
+
+if [ -z "$dontUseNinjaCheck" -a -z "$checkPhase" ]; then
+    checkPhase=ninjaCheckPhase
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: I think we'd have to remove the meson check phase to avoid confusion. It would probably be whichever one is set first.

**Comment by jtojnar** [comment]
One thing `meson test` has, it accepts extra arguments. For example, I can filter the tests or wrap them in `xvfb-run`: http://mesonbuild.com/Unit-tests.html#testing-tool

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh (hunk 2 above)
Comment (orivej) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: It looks like no package is using `buildFlags` with ninja (and the only package using `ninjaFlags` is `otfcc`); it seems counterproductive to support `installFlags`. I'd rather remove `buildFlags` from ninja flags array, since it is not needed and it is documented as a `make` flags variable.

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh [hunk 5]
```diff
@@ -48,3 +45,40 @@ ninjaInstallPhase() {
 if [ -z "$dontUseNinjaInstall" -a -z "$installPhase" ]; then
     installPhase=ninjaInstallPhase
 fi
+
+ninjaCheckPhase() {
+    runHook preCheck
+
+    if [ -z "${checkTarget:-}" ]; then
+        if ninja -n test >/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: Thanks!

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh (hunk 2 above)
Comment (matthewbauer) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: My opinion here is that we should reuse as much of the generic builder as possible. There are definitely some uses for this (for instance with checkFlags below). I can change the doc to reflect that {build|install|check}Flags passes flags to whatever build system you are using - not just make. As meson & cmake replace autotools, I think it is important we make them just as poewrful.

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (orivej) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: There is no change is behaviour, but there is also no need to change `-l$NIX_BUILD_CORES` to `-l$buildCores`. I don't want to see `-l1` in the command line because `-l` should be equal to the total number of busy cores across all concurrent builds; it's value should not be affected by the level of parallelism of a particular build.

**Comment by matthewbauer** [comment]
@jtojnar That is definitely a disadvantage of doing ```make check```. But, I believe ```ninja test``` will pass these correctly. Ideally you should just have to pass them in as ```checkFlags = "...";```.

**Comment by orivej** [comment]
Ninja does not have flags, it has only targets.

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (orivej) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: Since `meson` depends on `ninja`, `meson` hook will run first and set the `checkPhase`.

[the remark this answers is not in the document]
Comment (orivej) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: Oops, @Ericson2314 was working on it, but currently this is not the case, and the order in `nativeBuildInputs` is decisive.

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh [hunk 6]
```diff
@@ -1,26 +1,21 @@
...
+        $ninjaFlags "${ninjaFlagsArray[@]}"
+    )
 
-        echoCmd 'build flags' "${flagsArray[@]}"
-        ninja "${flagsArray[@]}"
-        unset flagsArray
-    fi
+    echoCmd 'build flags' "${flagsArray[@]}"
+    ninja "${flagsArray[@]}"
+    unset flagsArray
```
Comment (orivej) [inline pkgs/development/tools/build-managers/ninja/setup-hook.sh]: I have just noticed: since `flagsArray` is local, we don't have to unset it.

**Review by orivej — COMMENTED** [review]
I can't reproduce ofborg failure…

**Comment by jtojnar** [comment]
As a side note, it would be nice to have support for wrapping tests in the setup hooks so that we do not need to do

https://github.com/NixOS/nixpkgs/blob/0925c482c8378e8a6a8724edc6596ea60f90a113/pkgs/development/libraries/libdazzle/default.nix#L28-L33

**Comment by matthewbauer** [comment]
> As a side note, it would be nice to have support for wrapping tests in the setup hooks so that we do not need to do
> 
> [nixpkgs/pkgs/development/libraries/libdazzle/default.nix](https://github.com/NixOS/nixpkgs/blob/0925c482c8378e8a6a8724edc6596ea60f90a113/pkgs/development/libraries/libdazzle/default.nix#L28-L33)
> Lines 28 to 33 in [0925c48](/NixOS/nixpkgs/commit/0925c482c8378e8a6a8724edc6596ea60f90a113)
>    checkPhase = '' 
>      export NO_AT_BRIDGE=1 
>      xvfb-run -s '-screen 0 800x600x24' dbus-run-session \ 
>        --config-file=${dbus.daemon}/share/dbus-1/session.conf \ 
>        meson test --print-errorlogs 
>    '';

Yes! This would be nice to abstract out. It would also be useful with my "exeWrapper" work:

#50212 

So, you could run tests after cross compiling. It gets a little bit complicated because you may not have all of the right tools cross built in this case, but it's a similar principle to dbus-run-session.

File: pkgs/development/libraries/cmark/default.nix [hunk 7]
```diff
@@ -13,9 +13,9 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [ cmake ];
   doCheck = !stdenv.isDarwin;
-  checkPhase = ''
+  preCheck = ''
     export LD_LIBRARY_PATH=$(readlink -f ./src)
-    CTEST_OUTPUT_ON_FAILURE=1 make test
+    export CTEST_OUTPUT_ON_FAILURE=1
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/development/libraries/cmark/default.nix]: Yes. Also maybe CTEST_PARALLEL_LEVEL?

File: pkgs/development/tools/build-managers/ninja/setup-hook.sh (hunk 6 above)
Comment (matthewbauer) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: I noticed that too but thought it might be to avoid it leaking into the post hook?

Comment (orivej) [reply pkgs/development/tools/build-managers/ninja/setup-hook.sh]: `runHook` is a function, it has a new scope without local variables of the current function.

**Comment by matthewbauer** [comment]
@grahamcofborg build cmark at-spi2-atk

**Review by knedlsepp — COMMENTED** [review]
A little late to the party, but I think this might not work for cmake without ninja as we don't set checkTarget to "test" there (yet).
